client/service/src/config.rs

```diff
@@ -251,3 +253,36 @@ impl std::convert::From<PathBuf> for BasePath {
 		BasePath::new(path)
 	}
 }
+
+/// Callable object that execute tasks.
+#[derive(Clone)]
+pub struct TaskExecutor(Arc<dyn Fn(Pin<Box<dyn Future<Output = ()> + Send>>, TaskType) + Send + Sync>);
+
+impl std::fmt::Debug for TaskExecutor {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+		write!(f, "TaskExecutor")
+	}
+}
+
+impl std::convert::From<Arc<dyn Fn(Pin<Box<dyn Future<Output = ()> + Send>>, TaskType) + Send + Sync>>
+for TaskExecutor {
+	fn from(x: Arc<dyn Fn(Pin<Box<dyn Future<Output = ()> + Send>>, TaskType) + Send + Sync>)
+	-> Self {
+		Self(x)
+	}
+}
+
+impl std::ops::Deref for TaskExecutor {
+	type Target = Arc<dyn Fn(Pin<Box<dyn Future<Output = ()> + Send>>, TaskType) + Send + Sync>;
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
```

Comment:
This basically makes the struct callable. I'm not sure if this is the best solution.

1. We could say it's not callable at all and use a function instead
2. We could implement Fn but I think this is only available in nightly
3. It might be useful to be able to get the `Arc<Fn(future)>` from TaskExecutor

Any advice on what would be the cleanest implementation?

Reply:
As all the `TaskExecutor` has to do is to spawn futures and be cloned, I think the best idea would be to hide the internals and just add a spawn method.

Reply:
While `Arc<dyn Fn(Pin<Box<dyn Future<Output = ()> + Send>>, TaskType) + Send + Sync>` is not the most the most elegant type, anyone who knows the Rust programming language can understand this type. If you put it in a new `TaskExecutor` struct, now everyone has to go look at the definition of this `TaskExecutor` to know how to use it.

I know it's not much, but having so many typedefs and wrapping structs is in my opinion a major pain point in the Substrate code base as whole.

Reply:
Oh well. My goal was not to wrap it, I just needed to implement Debug and there are only 3 possible ways:

1. Wrapping that type and derive Debug on the whole Configuration
2. Writing a complete Debug impl for Debug and handle special cases separately (more maintenance)
3. Use a crate that would handle that kind of use case (not sure which one) which will reduce the maintenance but add a dependency

I think the trade-off is fine here especially with the `From` because you can simply pass a closure and do `.into()`. I think it would be better if I document it better so anyone who bumps into this will easily know what to do.

Reply:
(The other good point is that you don't need to import Arc and Pin, not even TaskExecutor, you just pass the closure and use `.into()`)

Reply:
done

Reply:
@expenses done (I just noticed your comment now)